baseapp/baseapp.go

```diff
@@ -623,16 +646,32 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		return err.Result()
 	}
 
-	// run the ante handler
+	// Execute the ante handler if one is defined. Note that during deliverTx, if
+	// execution fails and thus needs to abort, state modifications will not be
+	// persisted due to the transient CacheWrap multi-store that is used. Otherwise,
+	// during checkTx, state modifications are persisted between respective checkTx
+	// calls.
 	if app.anteHandler != nil {
-		newCtx, result, abort := app.anteHandler(ctx, tx, (mode == runTxModeSimulate))
+		var msCache sdk.CacheMultiStore
+
+		anteCtx := ctx
+		if mode == runTxModeDeliver {
+			anteCtx, msCache = app.cacheTxContext(ctx, txBytes, mode)
+		}
+
+		newCtx, result, abort := app.anteHandler(anteCtx, tx, (mode == runTxModeSimulate))
 		if abort {
 			return result
 		}
+
 		if !newCtx.IsZero() {
 			ctx = newCtx
 		}
 
+		if mode == runTxModeDeliver && msCache != nil {
```

Comment:
When would `msCache` be `nil`?

Reply:
It would never, just being safe here. But since we also don't check it when writing after running messages, I'll remove it.